### ISSUE TITLE
Implementing simple cooperative threads in C - Stephen Brennan

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6750]
Implementing simple cooperative threads in C - Stephen Brennan<br>
<br>
https://ift.tt/2LXdTxf<br>
<br>
Stephen Brennan<br>
<br>
<div>
<img title="Implementing simple cooperative threads in C &bull; Stephen Brennan" src="https://brennan.io/images/me.jpg"><br>
<div>
<div>
<h1>Implementing simple cooperative threads in C</h1>
<em>Stephen Brennan &bull; 24 May 2020</em>
<p>Multitasking, like many services an operating system provides, is something we take for granted so much that it can feel mundane. With our powerful smartphones and computers, the idea of a computer <em>not</em> being able to juggle hundreds of processes feels alien. I think it&rsquo;s features like this that make computers incredibly useful, but also make them feel so complicated and magical.</p>
<p>It&rsquo;s hard to play around with the code that implements multitasking, and it&rsquo;s not obvious how to implement it yourself without building a whole OS. I&rsquo;m a firm believer that you don&rsquo;t truly understand something until you&rsquo;ve implemented it yourself, so I wanted to write an article that lets people play around with a simple thread implementation. In this post, we&rsquo;ll implement simple threads in a normal C program (not an operating system).</p>
<h2>Detour through setjmp and longjmp</h2>
<p>This scheduler is going to rely heavily on the functions <code>setjmp()</code> and <code>longjmp()</code>. They feel a bit magical, so I want to first describe what they do, and spend a little time demystifying how they do it.</p>
<p>The function <code>setjmp()</code> is a way of recording information about where a program is in its execution, so that you can later jump back to that point. You give it a variable of type <code>jmp_buf</code>, in which it will store that information. <code>setjmp()</code> returns 0 the first time it returns.</p>
<p>Later on, you can use the function <code>longjmp(jmp_buf, value)</code> to immediately begin execution back at the point where you called <code>setjmp()</code>. To your program, it will look like <code>setjmp()</code> returned a <em>second time.</em> The <code>value</code> argument you pass to <code>longjmp()</code> will be returned this time, to help differentiate the second return. Here&rsquo;s an example to help illustrate:</p>
<div>
<div>
<pre><code><span>#include &lt;stdio.h&gt; #include &lt;setjmp.h&gt; </span> <span>jmp_buf</span> <span>saved_location</span><span>;</span> <span>int</span> <span>main</span><span>(</span><span>int</span> <span>argc</span><span>,</span> <span>char</span> <span>**</span><span>argv</span><span>)</span> <span>{</span> <span>if</span> <span>(</span><span>setjmp</span><span>(</span><span>saved_location</span><span>)</span> <span>==</span> <span>0</span><span>)</span> <span>{</span> <span>printf</span><span>(</span><span>"We have successfully set up our jump buffer!</span><span>\n</span><span>"</span><span>);</span> <span>}</span> <span>else</span> <span>{</span> <span>printf</span><span>(</span><span>"We jumped!</span><span>\n</span><span>"</span><span>);</span> <span>return</span> <span>0</span><span>;</span> <span>}</span> <span>printf</span><span>(</span><span>"Getting ready to jump!</span><span>\n</span><span>"</span><span>);</span> <span>longjmp</span><span>(</span><span>saved_location</span><span>,</span> <span>1</span><span>);</span> <span>printf</span><span>(</span><span>"This will never execute...</span><span>\n</span><span>"</span><span>);</span> <span>return</span> <span>0</span><span>;</span> <span>}</span> </code></pre>
</div>
</div>
<p>If you compile and run this program, you will get the following output:</p>
<div>
<div>
<pre><code>We have successfully set up our jump buffer! Getting ready to jump! We jumped! </code></pre>
</div>
</div>
<p>Wild! It&rsquo;s like a goto statement, but it can even be used to jump outside of a function. It&rsquo;s also a lot more difficult to read than a goto, since it looks like a regular function call. If your code used <code>setjmp()</code> and <code>longjmp()</code> liberally, it would be incredibly confusing for anyone (including yourself) to read.</p>
<p>Like with goto, the common advice is to avoid <code>setjmp()</code> and <code>longjmp()</code>. However, just like with goto, there are some times where it can be useful to use sparingly, and in a consistent way. A scheduler needs to be able to switch contexts, and so we&rsquo;ll have to use these functions responsibly. Most importantly, we&rsquo;ll hide the use of these functions from our API, so that users of our scheduler won&rsquo;t have to deal with that kind of complexity.</p>
<h2>Setjmp and longjmp don&rsquo;t save your stack</h2>
<p>The <code>setjmp()</code> and <code>longjmp()</code> functions aren&rsquo;t designed to support just <em>any</em> kind of jumping around, however. They&rsquo;re designed for a pretty particular use case. Imagine that you are doing something complicated, like making an HTTP request. This will involve a complicated set of function calls, and if any of them fail, you&rsquo;ll need to return a special error code from each one of them. This leads to code like this, everywhere you call a function (possibly dozens of times):</p>
<div>
<div>
<pre><code><span>int</span> <span>rv</span> <span>=</span> <span>do_function_call</span><span>();</span> <span>if</span> <span>(</span><span>rv</span> <span>!=</span> <span>SUCCESS</span><span>)</span> <span>{</span> <span>return</span> <span>rv</span><span>;</span> <span>}</span> </code></pre>
</div>
</div>
<p>The idea of <code>setjmp()</code> and <code>longjmp()</code> is that you can use <code>setjmp()</code> to save your place just before starting something complex. Then, you could centralize all of your error handling into one place:</p>
<div>
<div>
<pre><code><span>int</span> <span>rv</span><span>;</span> <span>jmp_buf</span> <span>buf</span><span>;</span> <span>if</span> <span>((</span><span>rv</span> <span>=</span> <span>setjmp</span><span>(</span><span>buf</span><span>))</span> <span>!=</span> <span>0</span><span>)</span> <span>{</span> <span>/* handle errors here */</span> <span>return</span><span>;</span> <span>}</span> <span>do_complicated_task</span><span>(</span><span>buf</span><span>,</span> <span>args</span><span>...);</span> </code></pre>
</div>
</div>
<p>If any function involved in <code>do_complicated_task()</code> fails, it would just <code>longjmp(buf, error_code)</code>. This means that every function within <code>do_complicated_task()</code> can assume that every function call is a success, which means you can get rid of that error handling code for each function call. (In practice, this is almost never done, but that&rsquo;s a separate blog post.)</p>
<p>The big idea here is that <code>longjmp()</code> only allows you to jump <em>out</em> of deeply nested functions. You can&rsquo;t jump back <em>into</em> a deeply nested function which you had formerly jumped out of. Here&rsquo;s an illustration of the stack when you jump out of a function. The asterisk <code>(*)</code> marks the stack pointer which <code>setjmp()</code> stored.</p>
<div>
<div>
<pre><code> | Stack before longjmp | Stack after longjmp +-------------------------+---------------------------- stack | main() (*) | main() grows | do_http_request() | down | send_a_header() | | | write_bytes() | v | write() - fails! | </code></pre>
</div>
</div>
<p>You can see that we only move back up the stack, and so there is no risk of data corruption. On the other hand, imagine if you wanted to jump between tasks. If you call <code>setjmp()</code> and then return, do some other stuff, and then attempt to resume what you were doing before, you&rsquo;ll have a problem:</p>
<div>
<div>
<pre><code> | Stack at setjmp() | Stack later | Stack after longjmp() +-------------------+------------------+---------------------- stack | main() | main() | main() grows | do_task_one() | do_task_two() | do_stack_two() down | subtask() | subtask() | subtask() | | foo() | | ??? v | bar() (*) | (*) | ??? (*) </code></pre>
</div>
</div>
<p>The stack pointer which <code>setjmp()</code> saved will point at a stack frame which no longer exists, and may have been overwritten at some point with other data. When you try to <code>longjmp()</code> back into the function you have already returned from, you&rsquo;ll start experiencing some really weird behavior that will probably crash your program.</p>
<p>The moral of this story is that, if you want to use setjmp() and longjmp() to jump between complex tasks like this, you need to make sure each task has <em>its own separate stack.</em> This completely eliminates the problem, because when <code>longjmp()</code> resets the stack pointer, it will swap stacks for you, and no stack overwriting will take place.</p>
<h2>Let&rsquo;s make a scheduler API</h2>
<p>That was a bit of a long diversion, but equipped with this knowledge, we should be able to implement userspace threads. To start out, I found it quite helpful to design the API which should be used to initialize, create, and run the threads. Doing this ahead of time really helps understand what we&rsquo;re trying to build!</p>
<div>
<div>
<pre><code><span>void</span> <span>scheduler_init</span><span>(</span><span>void</span><span>);</span> <span>void</span> <span>scheduler_create_task</span><span>(</span><span>void</span> <span>(</span><span>*</span><span>func</span><span>)(</span><span>void</span><span>*</span><span>),</span> <span>void</span> <span>*</span><span>arg</span><span>);</span> <span>void</span> <span>scheduler_run</span><span>(</span><span>void</span><span>);</span> </code></pre>
</div>
</div>
<p>These functions will be used to initialize the scheduler, add tasks, and then eventually begin running tasks in the scheduler. Once we start <code>scheduler_run()</code>, it will run until all tasks are completed. For tasks which are running, they will have the following APIs:</p>
<div>
<div>
<pre><code><span>void</span> <span>scheduler_exit_current_task</span><span>(</span><span>void</span><span>);</span> <span>void</span> <span>scheduler_relinquish</span><span>(</span><span>void</span><span>);</span> </code></pre>
</div>
</div>
<p>The first function will exit the task. A task could also exit by returning from its function, so this is simply a convenience. The second function is how our threads will tell the scheduler to let another task run for a bit. When a task calls <code>scheduler_relinquish()</code>, it could be suspended a bit for a bit, while other tasks run, but eventually the function will return and the task can continue running.</p>
<p>To give a concrete example of the API, here&rsquo;s a hypothetical use of this API, which we&rsquo;ll use to test our scheduler:</p>
<div>
<div>
<pre><code><span>#include &lt;stdlib.h&gt; #include &lt;stdio.h&gt; </span> <span>#include "scheduler.h" </span> <span>struct</span> <span>tester_args</span> <span>{</span> <span>char</span> <span>*</span><span>name</span><span>;</span> <span>int</span> <span>iters</span><span>;</span> <span>};</span> <span>void</span> <span>tester</span><span>(</span><span>void</span> <span>*</span><span>arg</span><span>)</span> <span>{</span> <span>int</span> <span>i</span><span>;</span> <span>struct</span> <span>tester_args</span> <span>*</span><span>ta</span> <span>=</span> <span>(</span><span>struct</span> <span>tester_args</span> <span>*</span><span>)</span><span>arg</span><span>;</span> <span>for</span> <span>(</span><span>i</span> <span>=</span> <span>0</span><span>;</span> <span>i</span> <span>&lt;</span> <span>ta</span><span>-&gt;</span><span>iters</span><span>;</span> <span>i</span><span>++</span><span>)</span> <span>{</span> <span>printf</span><span>(</span><span>"task %s: %d</span><span>\n</span><span>"</span><span>,</span> <span>ta</span><span>-&gt;</span><span>name</span><span>,</span> <span>i</span><span>);</span> <span>scheduler_relinquish</span><span>();</span> <span>}</span> <span>free</span><span>(</span><span>ta</span><span>);</span> <span>}</span> <span>void</span> <span>create_test_task</span><span>(</span><span>char</span> <span>*</span><span>name</span><span>,</span> <span>int</span> <span>iters</span><span>)</span> <span>{</span> <span>struct</span> <span>tester_args</span> <span>*</span><span>ta</span> <span>=</span> <span>malloc</span><span>(</span><span>sizeof</span><span>(</span><span>*</span><span>ta</span><span>));</span> <span>ta</span><span>-&gt;</span><span>name</span> <span>=</span> <span>name</span><span>;</span> <span>ta</span><span>-&gt;</span><span>iters</span> <span>=</span> <span>iters</span><span>;</span> <span>scheduler_create_task</span><span>(</span><span>tester</span><span>,</span> <span>ta</span><span>);</span> <span>}</span> <span>int</span> <span>main</span><span>(</span><span>int</span> <span>argc</span><span>,</span> <span>char</span> <span>**</span><span>argv</span><span>)</span> <span>{</span> <span>scheduler_init</span><span>();</span> <span>create_test_task</span><span>(</span><span>"first"</span><span>,</span> <span>5</span><span>);</span> <span>create_test_task</span><span>(</span><span>"second"</span><span>,</span> <span>2</span><span>);</span> <span>scheduler_run</span><span>();</span> <span>printf</span><span>(</span><span>"Finished running all tasks!</span><span>\n</span><span>"</span><span>);</span> <span>return</span> <span>EXIT_SUCCESS</span><span>;</span> <span>}</span> </code></pre>
</div>
</div>
<p>In this example, we create two tasks which run the same function, but they&rsquo;ll use different arguments so that we can trace their execution separately. Each task iterates a set number of times. Each iteration, it prints out a message and then lets another task run. We would expect to see something like this as the output of this program:</p>
<div>
<div>
<pre><code>task first: 0 task second: 0 task first: 1 task second: 1 task first: 2 task first: 3 task first: 4 Finished running all tasks! </code></pre>
</div>
</div>
<h2>Let&rsquo;s implement the scheduler API</h2>
<p>To implement this API, we&rsquo;ll need some sort of internal representation of a task, so let&rsquo;s go ahead and put together fields we&rsquo;ll need:</p>
<div>
<div>
<pre><code><span>struct</span> <span>task</span> <span>{</span> <span>enum</span> <span>{</span> <span>ST_CREATED</span><span>,</span> <span>ST_RUNNING</span><span>,</span> <span>ST_WAITING</span><span>,</span> <span>}</span> <span>status</span><span>;</span> <span>int</span> <span>id</span><span>;</span> <span>jmp_buf</span> <span>buf</span><span>;</span> <span>void</span> <span>(</span><span>*</span><span>func</span><span>)(</span><span>void</span><span>*</span><span>);</span> <span>void</span> <span>*</span><span>arg</span><span>;</span> <span>struct</span> <span>sc_list_head</span> <span>task_list</span><span>;</span> <span>void</span> <span>*</span><span>stack_bottom</span><span>;</span> <span>void</span> <span>*</span><span>stack_top</span><span>;</span> <span>int</span> <span>stack_size</span><span>;</span> <span>};</span> </code></pre>
</div>
</div>
<p>Let&rsquo;s go through the fields one by one. All tasks should be in the &ldquo;created&rdquo; state as soon as they&rsquo;re created. Once a task starts executing, it will be in the &ldquo;running&rdquo; status, and if a task ever needed to wait for some asynchronous operation, it could be placed in the &ldquo;waiting&rdquo; state. The <code>id</code> field is just a unique identifier for the task. <code>buf</code> contains the data for when we <code>longjmp()</code> to resume the task. <code>func</code> and <code>arg</code> are passed to <code>scheduler_create_task()</code> and are necessary for starting the task. The <code>task_list</code> field is necessary to implement a doubly linked list of all tasks. The <code>stack_bottom</code>, <code>stack_top</code>, and <code>stack_size</code> fields all relate to the separate stack allocated for this task. The &ldquo;bottom&rdquo; is the address returned by <code>malloc()</code>, but the &ldquo;top&rdquo; is a pointer to the address directly above the region of memory. Since the x86 stack grows downward, we will need to set the stack pointer to <code>stack_top</code> rather than <code>stack_bottom</code>.</p>
<p>Given this, we can implement the <code>scheduler_create_task()</code> function:</p>
<div>
<div>
<pre><code><span>void</span> <span>scheduler_create_task</span><span>(</span><span>void</span> <span>(</span><span>*</span><span>func</span><span>)(</span><span>void</span> <span>*</span><span>),</span> <span>void</span> <span>*</span><span>arg</span><span>)</span> <span>{</span> <span>static</span> <span>int</span> <span>id</span> <span>=</span> <span>1</span><span>;</span> <span>struct</span> <span>task</span> <span>*</span><span>task</span> <span>=</span> <span>malloc</span><span>(</span><span>sizeof</span><span>(</span><span>*</span><span>task</span><span>));</span> <span>task</span><span>-&gt;</span><span>status</span> <span>=</span> <span>ST_CREATED</span><span>;</span> <span>task</span><span>-&gt;</span><span>func</span> <span>=</span> <span>func</span><span>;</span> <span>task</span><span>-&gt;</span><span>arg</span> <span>=</span> <span>arg</span><span>;</span> <span>task</span><span>-&gt;</span><span>id</span> <span>=</span> <span>id</span><span>++</span><span>;</span> <span>task</span><span>-&gt;</span><span>stack_size</span> <span>=</span> <span>16</span> <span>*</span> <span>1024</span><span>;</span> <span>task</span><span>-&gt;</span><span>stack_bottom</span> <span>=</span> <span>malloc</span><span>(</span><span>task</span><span>-&gt;</span><span>stack_size</span><span>);</span> <span>task</span><span>-&gt;</span><span>stack_top</span> <span>=</span> <span>task</span><span>-&gt;</span><span>stack_bottom</span> <span>+</span> <span>task</span><span>-&gt;</span><span>stack_size</span><span>;</span> <span>sc_list_insert_end</span><span>(</span><span>&amp;</span><span>priv</span><span>.</span><span>task_list</span><span>,</span> <span>&amp;</span><span>task</span><span>-&gt;</span><span>task_list</span><span>);</span> <span>}</span> </code></pre>
</div>
</div>
<p>Using a <code>static int</code> ensures that each time the function is called, the <code>id</code> field increments to a new number. Everything else should be self-explanatory, except the <code>sc_list_insert_end()</code>, which simply adds the <code>struct task</code> to the global list. The global list is stored within a second structure, which contains all the private scheduler data. This structure is presented below, along with its initialization function:</p>
<div>
<div>
<pre><code><span>struct</span> <span>scheduler_private</span> <span>{</span> <span>jmp_buf</span> <span>buf</span><span>;</span> <span>struct</span> <span>task</span> <span>*</span><span>current</span><span>;</span> <span>struct</span> <span>sc_list_head</span> <span>task_list</span><span>;</span> <span>}</span> <span>priv</span><span>;</span> <span>void</span> <span>scheduler_init</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>priv</span><span>.</span><span>current</span> <span>=</span> <span>NULL</span><span>;</span> <span>sc_list_init</span><span>(</span><span>&amp;</span><span>priv</span><span>.</span><span>task_list</span><span>);</span> <span>}</span> </code></pre>
</div>
</div>
<p>The <code>task_list</code> field is used to refer to the list of tasks (unsurprisingly). The <code>current</code> field is used to store the currently executing task (or null if none is curently running). Most importantly, the <code>buf</code> field will be used to jump into the code of <code>scheduler_run()</code>:</p>
<div>
<div>
<pre><code><span>enum</span> <span>{</span> <span>INIT</span><span>=</span><span>0</span><span>,</span> <span>SCHEDULE</span><span>,</span> <span>EXIT_TASK</span><span>,</span> <span>};</span> <span>void</span> <span>scheduler_run</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>/* This is the exit path for the scheduler! */</span> <span>switch</span> <span>(</span><span>setjmp</span><span>(</span><span>priv</span><span>.</span><span>buf</span><span>))</span> <span>{</span> <span>case</span> <span>EXIT_TASK</span><span>:</span> <span>scheduler_free_current_task</span><span>();</span> <span>case</span> <span>INIT</span><span>:</span> <span>case</span> <span>SCHEDULE</span><span>:</span> <span>schedule</span><span>();</span> <span>/* if return, there's nothing else to do and we exit */</span> <span>return</span><span>;</span> <span>default:</span> <span>fprintf</span><span>(</span><span>stderr</span><span>,</span> <span>"Uh oh, scheduler error</span><span>\n</span><span>"</span><span>);</span> <span>return</span><span>;</span> <span>}</span> <span>}</span> </code></pre>
</div>
</div>
<p>As soon as the <code>scheduler_run()</code> function is called, we set the <code>setjmp()</code> buffer so we can always return to this function. The first time, 0 (<code>INIT</code>) is returned, and we immediately call <code>schedule()</code>. Subsequently, we can pass the <code>SCHEDULE</code> or <code>EXIT_TASK</code> constants into <code>longjmp()</code>, which will trigger different behaviors. Let&rsquo;s ignore the <code>EXIT_TASK</code> case for now, and go directly into the implementation of <code>schedule()</code>:</p>
<div>
<div>
<pre><code><span>static</span> <span>void</span> <span>schedule</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>struct</span> <span>task</span> <span>*</span><span>next</span> <span>=</span> <span>scheduler_choose_task</span><span>();</span> <span>if</span> <span>(</span><span>!</span><span>next</span><span>)</span> <span>{</span> <span>return</span><span>;</span> <span>}</span> <span>priv</span><span>.</span><span>current</span> <span>=</span> <span>next</span><span>;</span> <span>if</span> <span>(</span><span>next</span><span>-&gt;</span><span>status</span> <span>==</span> <span>ST_CREATED</span><span>)</span> <span>{</span> <span>/* * This task has not been started yet. Assign a new stack * pointer, run the task, and exit it at the end. */</span> <span>register</span> <span>void</span> <span>*</span><span>top</span> <span>=</span> <span>next</span><span>-&gt;</span><span>stack_top</span><span>;</span> <span>asm</span> <span>volatile</span><span>(</span> <span>"mov %[rs], %%rsp </span><span>\n</span><span>"</span> <span>:</span> <span>[</span> <span>rs</span> <span>]</span> <span>"+r"</span> <span>(</span><span>top</span><span>)</span> <span>::</span> <span>);</span> <span>/* * Run the task function */</span> <span>next</span><span>-&gt;</span><span>status</span> <span>=</span> <span>ST_RUNNING</span><span>;</span> <span>next</span><span>-&gt;</span><span>func</span><span>(</span><span>next</span><span>-&gt;</span><span>arg</span><span>);</span> <span>/* * The stack pointer should be back where we set it. Returning would be * a very, very bad idea. Let's instead exit */</span> <span>scheduler_exit_current_task</span><span>();</span> <span>}</span> <span>else</span> <span>{</span> <span>longjmp</span><span>(</span><span>next</span><span>-&gt;</span><span>buf</span><span>,</span> <span>1</span><span>);</span> <span>}</span> <span>/* NO RETURN */</span> <span>}</span> </code></pre>
</div>
</div>
<p>First, we call an internal function to select the next task which should be run. This is going to be a simple round-robin scheduler, so it just chooses the next ready task in the task list. If this function returned NULL, then we have no more tasks to run, and we return. Otherwise, we need to either start the task running (if it is in the <code>ST_CREATED</code> state) or resume running it.</p>
<p>To start a created task, we use an x86_64 assembly instruction to assign the <code>stack_top</code> field to the <code>rsp</code> register (stack pointer). Then we change the task state, run the function, and exit if the function returns. Note that <code>setjmp()</code> and <code>longjmp()</code> store and swap stack pointers, so this is the only time where we&rsquo;ll need to use assembly to modify the stack pointer.</p>
<p>If the task has already been started, then the <code>buf</code> field should contain the context we need to <code>longjmp()</code> into to resume the task, so we just do that. Next, let&rsquo;s look at the helper function which selects the next task to run. This is the heart of a scheduler, and like I said earlier, this is a round-robin scheduler:</p>
<div>
<div>
<pre><code><span>static</span> <span>struct</span> <span>task</span> <span>*</span><span>scheduler_choose_task</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>struct</span> <span>task</span> <span>*</span><span>task</span><span>;</span> <span>sc_list_for_each_entry</span><span>(</span><span>task</span><span>,</span> <span>&amp;</span><span>priv</span><span>.</span><span>task_list</span><span>,</span> <span>task_list</span><span>,</span> <span>struct</span> <span>task</span><span>)</span> <span>{</span> <span>if</span> <span>(</span><span>task</span><span>-&gt;</span><span>status</span> <span>==</span> <span>ST_RUNNING</span> <span>||</span> <span>task</span><span>-&gt;</span><span>status</span> <span>==</span> <span>ST_CREATED</span><span>)</span> <span>{</span> <span>sc_list_remove</span><span>(</span><span>&amp;</span><span>task</span><span>-&gt;</span><span>task_list</span><span>);</span> <span>sc_list_insert_end</span><span>(</span><span>&amp;</span><span>priv</span><span>.</span><span>task_list</span><span>,</span> <span>&amp;</span><span>task</span><span>-&gt;</span><span>task_list</span><span>);</span> <span>return</span> <span>task</span><span>;</span> <span>}</span> <span>}</span> <span>return</span> <span>NULL</span><span>;</span> <span>}</span> </code></pre>
</div>
</div>
<p>If you&rsquo;re unfamiliar with my linked list implementation (which is taken from the Linux kernel), that&rsquo;s ok. The <code>sc_list_for_each_entry()</code> function is a macro that lets us iterate over each task in the task list. The first eligible (not waiting) task we find is removed from its current position and inserted at the end of the task list. This ensures that next time we run the scheduler, we&rsquo;ll get a different task (if there is another). We return this first eligible task, or NULL if there were no tasks at all.</p>
<p>Finally, let&rsquo;s get to the implementation of <code>scheduler_relinquish()</code> to see how a task can switch itself out:</p>
<div>
<div>
<pre><code><span>void</span> <span>scheduler_relinquish</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>if</span> <span>(</span><span>setjmp</span><span>(</span><span>priv</span><span>.</span><span>current</span><span>-&gt;</span><span>buf</span><span>))</span> <span>{</span> <span>return</span><span>;</span> <span>}</span> <span>else</span> <span>{</span> <span>longjmp</span><span>(</span><span>priv</span><span>.</span><span>buf</span><span>,</span> <span>SCHEDULE</span><span>);</span> <span>}</span> <span>}</span> </code></pre>
</div>
</div>
<p>This is the other use of the <code>setjmp()</code> function in our scheduler. As such it can be slightly confusing. When a task calls this function, we use <code>setjmp()</code> to save our current context (which includes the current stack pointer). Then, we use <code>longjmp()</code> to enter into the scheduler (back in <code>scheduler_run()</code>), and we pass the <code>SCHEDULE</code> function asking to schedule a new task.</p>
<p>When the task gets resumed, the <code>setjmp()</code> function will return non-zero and we&rsquo;ll return out to whatever the task was doing before!</p>
<p>Finally, here&rsquo;s what happens when a task exits (either by explicitly calling the exit function, or by returning from its task function):</p>
<div>
<div>
<pre><code><span>void</span> <span>scheduler_exit_current_task</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>struct</span> <span>task</span> <span>*</span><span>task</span> <span>=</span> <span>priv</span><span>.</span><span>current</span><span>;</span> <span>sc_list_remove</span><span>(</span><span>&amp;</span><span>task</span><span>-&gt;</span><span>task_list</span><span>);</span> <span>longjmp</span><span>(</span><span>priv</span><span>.</span><span>buf</span><span>,</span> <span>EXIT_TASK</span><span>);</span> <span>/* NO RETURN */</span> <span>}</span> <span>static</span> <span>void</span> <span>scheduler_free_current_task</span><span>(</span><span>void</span><span>)</span> <span>{</span> <span>struct</span> <span>task</span> <span>*</span><span>task</span> <span>=</span> <span>priv</span><span>.</span><span>current</span><span>;</span> <span>priv</span><span>.</span><span>current</span> <span>=</span> <span>NULL</span><span>;</span> <span>free</span><span>(</span><span>task</span><span>-&gt;</span><span>stack_bottom</span><span>);</span> <span>free</span><span>(</span><span>task</span><span>);</span> <span>}</span> </code></pre>
</div>
</div>
<p>This process comes in two parts: the first function is called directly by the task. We remove the task&rsquo;s entry from the task list, since it should no longer be scheduled. Then, we <code>longjmp()</code> back to the <code>scheduler_run()</code> function. This time, we use <code>EXIT_TASK</code>. This indicates to the scheduler that, before it schedules a new task, it should first call <code>scheduler_free_current_task()</code>. If you scroll back up to <code>scheduler_run()</code>, you&rsquo;ll see this is exactly what <code>scheduler_run()</code> does.</p>
<p>We have to do this in two parts because, when <code>scheduler_exit_current_task()</code> is called, it is actively using the stack contained in the task struct. If you free the stack while still using it, there&rsquo;s the chance that the function will still access the very stack memory we just freed! To ensure this doesn&rsquo;t happen, we have to <code>longjmp()</code> back to the scheduler, which is using a separate stack. Then we can safely free the task&rsquo;s data.</p>
<p>With that, we&rsquo;ve covered the entire implementation of this scheduler. If you were to go ahead and compile this, along with my linked list implementation and the main program above, you would have a working scheduler! Instead of all that copying and pasting, I&rsquo;d encourage you to check out the <a href="https://github.com/brenns10/userspace_cooperative_multitasking">github repository</a> which contains all this code.</p>
<h2>Why is this useful?</h2>
<p>If you&rsquo;ve gotten this far, I assume I don&rsquo;t need to convince you that this is interesting. However, it may not seem all that useful. After all, you can use &ldquo;real&rdquo; threads in C, which can run in parallel and don&rsquo;t need to wait for each other to call <code>scheduler_relinquish()</code>.</p>
<p>However, I see this as a jumping off point for a whole series of exciting implementations of useful features. For I/O heavy tasks, this could also be used to simply implement a single-threaded async application, the way that Python&rsquo;s new async utilities work. This system could also implement generators and coroutines. Finally, with enough effort, this system could even be coupled with &ldquo;real&rdquo; operating system threads to provide more parallelism where necessary. Each of these ideas is a fun project which I&rsquo;d encourage the reader to try before I get around to writing a new article about them!</p>
<h2>Is it safe?</h2>
<p>I mean, probably not! It&rsquo;s probably not safe to use inline assembly to modify the stack pointer. Don&rsquo;t use it in your production code, but do use it to mess around and explore!</p>
<p>A safer implementation of this system could be built on the &ldquo;ucontext&rdquo; API (see <code>man getcontext</code>), which provides a way to swap between these types of userspace &ldquo;threads&rdquo; without needing to meddle with inline assembly. Unfortunately, the API is non-standard (it was removed from the latest POSIX spec). However, you can still use this API, as it is part of <code>glibc</code>.</p>
<h2>How to make it preemptive?</h2>
<p>As it is currently written, this scheduler only works if threads explicitly hand off control back to the scheduler. This is bad for a general purpose implementation like an operating system, because a poorly behaved thread could prevent all the others from running. (Of course, that didn&rsquo;t stop MS-DOS from using cooperative multitasking!). I don&rsquo;t think that makes cooperative multitasking bad, it&rsquo;s just going to depend on the application.</p>
<p>If one used the non-standard &ldquo;ucontext&rdquo; API, then POSIX signals would actually store the context of the previously executing code. By setting a periodic timer signal, a userspace scheduler could actually get preemptive multitasking working! This is another really cool project that I hope to try out and write about soon.</p>
<p>If you&rsquo;ve gotten this far, thanks for reading, and I hope you get the chance to try out a fun project based on this!</p>
<hr>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Stephen Brennan's Blog<br>
<br>
May 25, 2020 at 05:25PM
